heartbeat/monitors/active/icmp/stdloop.go

```diff
@@ -0,0 +1,402 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package icmp
+
+import (
+	"bytes"
+	"encoding/binary"
+	"errors"
+	"fmt"
+	"math/rand"
+	"net"
+	"os"
+	"runtime"
+	"sync"
+	"time"
+
+	"golang.org/x/net/icmp"
+	"golang.org/x/net/ipv4"
+	"golang.org/x/net/ipv6"
+)
+
+type stdICMPLoop struct {
+	conn4, conn6 *icmp.PacketConn
+	recv         chan packet
+
+	mutex    sync.Mutex
+	requests map[requestID]*requestContext
+}
+
+type timeoutError struct {
+}
+
+const (
+	// iana types
+	protocolICMP     = 1
+	protocolIPv6ICMP = 58
+)
+
+type packet struct {
+	ts   time.Time
+	addr net.Addr
+
+	Type     icmp.Type // type, either ipv4.ICMPType or ipv6.ICMPType
+	Code     int       // code
+	Checksum int       // checksum
+	Echo     icmp.Echo
+}
+
+type requestID struct {
+	addr  string
+	proto int
+	id    int
+	seq   int
+}
+
+type requestContext struct {
+	l      *stdICMPLoop
+	id     requestID
+	ts     time.Time
+	result chan requestResult
+}
+
+type requestResult struct {
+	packet packet
+	err    error
+}
+
+// stdLoop is a singleton for our main ICMP loop since it doesn't
+// make sense to have multiples. While having a singleton is ugly
+// is mandatory for the ICMP interface in go, where all monitors
+// must share a single loop.
+// These vars should not be used directly, but rather getStdLoop
+// should be invoked to initialize and return stdLoop.
+var (
+	stdICMPLoopInit      sync.Once
+	stdICMPLoopSingleton *stdICMPLoop
+)
+
+func getStdLoop() (*stdICMPLoop, error) {
```

Comment:
This function is new, and cleanly abstracts the singleton nature of the vars above, which previously were accessed directly in `icmp.go`